MSC 1794 - Federation v2 Invite API (#1794)
matrix-org/matrix-doc: proposals/1794-federation-v2-invites.md

# MSC 1794 - Federation v2 Invite API

This proposal adds a new `/invite` API to federation that supports different
room versions.

## Motivation

It is planned for future room versions to be able to change the format of events
in various ways. To support this, all servers must know the room version
whenever they need to parse an event.  Currently the `/invite` API does not
include the room version, so the target server would be unable to parse the event included in the payload.

## Proposal

Add a new version of the invite API under the prefix `/_matrix/federation/v2`,
which has a payload of:

```
PUT /_matrix/federation/v2/invite/<room_id>/<event_id>

{
    "room_version": <room_version>,
    "event": { ... },
    "invite_room_state": [ ... ]
}
```

The differences between this and `v1` are:

1. The payload in `v1` is the event, while in `v2` the event is instead placed
   under an `"event"` key. This is for consistency with other APIs, and to allow
   extra data to be added to the request payload separately from the event.
2. A required field called `"room_version"` is added that specifies the room
   version.

> 3. The `"invite_room_state"` is moved from the `unsigned` section of the event

Reviewer: on retrospect, I don't know if this is a good idea. We spent a lot of time going through the spec to make sure the `invite_room_state` was correctly documented to be under the unsigned object. Given this is for s2s communications though, this is probably fine.

# MSC 1794 - Federation v2 Invite API

This proposal adds a new `/invite` API to federation that supports different
room versions.

## Motivation

It is planned for future room versions to be able to change the format of events

> in various ways, and so to support this all servers must know the room version

Reviewer 2: ```suggestion
in various ways. To support this, all servers must know the room version
```

whenever they need to parse an event.  Currently the `/invite` API does not

> include the room version, and so the event included in the payload would not be

Reviewer 2: ```suggestion
include the room version, so the target server would be unable to parse the event included in the payload.
```

> able to be parsed.

Reviewer 2: ```suggestion

```

## Proposal

Add a new version of the invite API under the prefix `/_matrix/federation/v2`,
which has a payload of:

```
PUT /_matrix/federation/v2/invite/<room_id>/<event_id>

{
    "room_version": <room_version>,
    "event": { ... },

>     "invite_room_state": { ... }

Reviewer 3: In the v1 API, `invite_room_state` is an array rather than an object.  Is that also the case here (and "{...}" should be "[...]")?

Author: Err, yes! Whoops

}
```

The differences between this and `v1` are:

1. The payload in `v1` is the event, while in `v2` the event is instead placed
   under an `"event"` key. This is for consistency with other APIs, and to allow
   extra data to be added to the request payload separately from the event.
2. A required field called `"room_version"` is added that specifies the room
   version.
3. The `"invite_room_state"` is moved from the `unsigned` section of the event
   to a top level key. The key `invite_room_state` being in the `event.unsigned`
   was a hack due to point 1. above.

The response is identical to `v1`, except that:

1. If the receiving server does not support the given room version the

>    appropriate incompatible-room-version exception is returned, in the same way

Reviewer 4: s/exception/error/?

Author: ```suggestion
   appropriate incompatible-room-version error is returned, in the same way
```

   as e.g. for `/make_join` APIs.
2. The response payload is no longer in the format of `[200, { ... }]`, and is
   instead simply the `{ ... }` portion. This fixes a historical accident to
   bring the invite API into line with the rest of the federation API.

If a call to `v2` `/invite` results in a unrecognised request exception **AND**
the room version is `1` or `2` then the sending server should retry the request
with the `v1` API.

## Alternatives

### Reusing V1 API

> One alternative is to add a `room_version` query string parameter to the `v1`

Reviewer 5: I'm struggling to see how this makes any difference? Why would an older server take any notice of an extra body parameter any more than a query string parameter, surely it's going to still try to parse it as V1 unless it's been updated to notice the `room_version` parameter, and at that point you might as well conform to `/make_join`.

Author: > I'm struggling to see how this makes any difference? Why would an older server take any notice of an extra body parameter any more than a query string parameter, surely it's going to still try to parse it as V1 unless it's been updated to notice the `room_version` parameter, ...

Exactly? And so the old server will try and parse the event and quite possibly explode

> ... and at that point you might as well conform to `/make_join`.

I don't understand what you mean by "conform"?

Reviewer 5: It's not a huge issue, but it's a little sad that `/make_join` has a `ver` query param for specifying room version, and this endpoint puts it inside the body. It would be nice for this endpoint and `/make_join` to specify room versions the same way.

So what I'm trying to understand is why is it fine for `/make_join` to use a query param but not this. If I understood your argument correctly, old servers are going to explode no matter where we specify it so we might as well make all endpoints do version specifying the same way?

Reviewer 5: Okay, I think that makes sense. Thanks for the explanation.
